internal/controllers/mic_reconciler.go

```diff
@@ -188,23 +188,29 @@ func (mrhi *micReconcilerHelperImpl) updateStatusByMBSC(ctx context.Context, mic
 	}
 
 	patchFrom := client.MergeFrom(micObj.DeepCopy())
-	for _, imageState := range mbsc.Status.ImagesStates {
-		imageSpec := mrhi.micHelper.GetModuleImageSpec(micObj, imageState.Image)
-		if imageSpec == nil {
-			// image not found in spec, ignore
+	for _, mbscImageState := range mbsc.Status.Images {
+		micImageSpec := mrhi.micHelper.GetModuleImageSpec(micObj, mbscImageState.Image)
+		if micImageSpec == nil {
+			// spec for image does not exists in MIC, meaning it was already deleted - ignore status
 			continue
 		}
-		switch imageState.Status {
-		case kmmv1beta1.ImageBuildFailed, kmmv1beta1.ImageSignFailed:
-			mrhi.micHelper.SetImageStatus(micObj, imageState.Image, kmmv1beta1.ImageDoesNotExist)
-		case kmmv1beta1.ImageBuildSucceeded:
-			if imageSpec.Sign != nil {
-				mrhi.micHelper.SetImageStatus(micObj, imageState.Image, kmmv1beta1.ImageNeedsSigning)
-			} else {
-				mrhi.micHelper.SetImageStatus(micObj, imageState.Image, kmmv1beta1.ImageExists)
-			}
-		case kmmv1beta1.ImageSignSucceeded:
-			mrhi.micHelper.SetImageStatus(micObj, imageState.Image, kmmv1beta1.ImageExists)
+
+		mbscStatus := mbscImageState.Status
+		mbscAction := mbscImageState.Action
+
+		switch {
+		case mbscStatus == kmmv1beta1.ActionFailure:
+			// any failure (build or sign) - image does not exists
+			mrhi.micHelper.SetImageStatus(micObj, mbscImageState.Image, kmmv1beta1.ImageDoesNotExist)
+		case mbscStatus == kmmv1beta1.ActionSuccess && mbscAction == kmmv1beta1.SignImage:
+			// sign action succeeded - image exists, nothing more to do
+			mrhi.micHelper.SetImageStatus(micObj, mbscImageState.Image, kmmv1beta1.ImageExists)
+		case mbscStatus == kmmv1beta1.ActionSuccess && micImageSpec.Sign != nil:
+			// build succeeded and sign exists - image needs to be signed
+			mrhi.micHelper.SetImageStatus(micObj, mbscImageState.Image, kmmv1beta1.ImageNeedsSigning)
+		case mbscStatus == kmmv1beta1.ActionSuccess:
+			// build succeeded, no sign - image exists
+			mrhi.micHelper.SetImageStatus(micObj, mbscImageState.Image, kmmv1beta1.ImageExists)
```

Comment:
```suggestion
switch {

case mbscStatus == kmmv1beta1.ActionFailure:
    // any failure (build or sign) - image does not exists
    mrhi.micHelper.SetImageStatus(micObj, mbscImageState.Image, kmmv1beta1.ImageDoesNotExist)

case mbscStatus == kmmv1beta1.ActionSuccess:

    if (mbscAction == kmmv1beta1.BuildImage && micImageSpec.Sign == nil) || mbscAction == kmmv1beta1.SignImage {
        // image exists, nothing more to do
        mrhi.micHelper.SetImageStatus(micObj, mbscImageState.Image, kmmv1beta1.ImageExists)
    } else if mbscAction == kmmv1beta1.BuildImage && micImageSpec.Sign != nil {
        // build succeeded and sign exists - image needs to be signed                                           
        mrhi.micHelper.SetImageStatus(micObj, mbscImageState.Image, kmmv1beta1.ImageNeedsSigning)
    }
```

I think it is easier to understand and to make sure we didn't forget any optional combination.

Reply:
Since github diff make it hard to read then this is the case:
```
switch {

case mbscStatus == kmmv1beta1.ActionFailure:
    // any failure (build or sign) - image does not exists
    mrhi.micHelper.SetImageStatus(micObj, mbscImageState.Image, kmmv1beta1.ImageDoesNotExist)

case mbscStatus == kmmv1beta1.ActionSuccess:

    if (mbscAction == kmmv1beta1.BuildImage && micImageSpec.Sign == nil) || mbscAction == kmmv1beta1.SignImage {
        // image exists, nothing more to do
        mrhi.micHelper.SetImageStatus(micObj, mbscImageState.Image, kmmv1beta1.ImageExists)
    } else if mbscAction == kmmv1beta1.BuildImage && micImageSpec.Sign != nil {
        // build succeeded and sign exists - image needs to be signed                                           
        mrhi.micHelper.SetImageStatus(micObj, mbscImageState.Image, kmmv1beta1.ImageNeedsSigning)
    }
}
```

Reply:
I wrote it especially this way in order to avoid using "switch" and "if", which in my opinion makes code a little bit harder to read. With only switch case we can see exactly what is going on